### PR TITLE
Introduce CirculatingSupplyUpdater

### PR DESCRIPTION
**Pull request by antooni**
This PR will add logic responsible for fetching circulating supplies from coingecko

Resolves L2B-2333

### CODE REVIEW

File: packages/backend/src/core/assets/CirculatingSupplyFormulaUpdater.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import { Clock } from '../Clock'
+import { PriceUpdater } from '../PriceUpdater'
+import { TaskQueue } from '../queue/TaskQueue'
+import { createSuppliedFormulaReports } from '../reports/createTotalSupplyFormulaReports'
+import { getTotalSupplyFormulaConfigHash } from '../reports/getTotalSupplyFormulaConfigHash'
+import { CirculatingSupplyUpdater } from '../totalSupply/CirculatingSupplyUpdater'
+import { AssetUpdater } from './AssetUpdater'
+
+// TODO: Make this class more generic
+// ProjectId, ChainId should be passed in the constructor
```
Comment (mateuszradomski) [inline packages/backend/src/core/assets/CirculatingSupplyFormulaUpdater.ts:26]: this comment can be removed

File: packages/backend/src/core/assets/CirculatingSupplyFormulaUpdater.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    private readonly logger: Logger,
+    private readonly minTimestamp: UnixTime,
+  ) {
+    assert(
+      tokens.every(
+        (token) =>
+          token.chainId === this.chainId &&
+          token.formula === 'circulatingSupply',
+      ),
+      'Programmer error: all tokens must be using circulatingSupply formula have the same chainId',
```
Comment (mateuszradomski) [inline packages/backend/src/core/assets/CirculatingSupplyFormulaUpdater.ts:50]:
```suggestion
      'Programmer error: all tokens must be using circulatingSupply formula and have the same chainId',
```

File: packages/backend/src/core/reports/createTotalSupplyFormulaReports.ts
Comment (mateuszradomski) [inline packages/backend/src/core/reports/createTotalSupplyFormulaReports.ts:1]: rename file to `createSuppliedFormulaReports`

File: packages/backend/src/core/totalSupply/CirculatingSupplyUpdater.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+        metricsId: CirculatingSupplyUpdater.name,
+      },
+    )
+    assert(
+      tokens.every(
+        (token) =>
+          token.chainId === this.chainId &&
+          token.formula === 'circulatingSupply',
+      ),
+      'Programmer error: all tokens must be using circulatingSupply formula have the same chainId',
```
Comment (mateuszradomski) [inline packages/backend/src/core/totalSupply/CirculatingSupplyUpdater.ts:47]:
```suggestion
      'Programmer error: all tokens must be using circulatingSupply formula and have the same chainId',
```

File: packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts [hunk 4]
```diff
@@ -29,33 +26,81 @@ export class CoingeckoQueryService {
...
+
+    const prices = (
+      await this.queryCoinMarketChartRange(
+        coingeckoId,
+        start,
+        end,
+        granularity,
+        address,
+      )
+    ).prices
```
Comment (mateuszradomski) [inline packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts:40]: why the braces `()`

File: packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts [hunk 5]
```diff
@@ -171,3 +228,19 @@ export function generateRangesToCallHourly(from: UnixTime, to: UnixTime) {
   }
   return ranges
 }
+
+export function approximateCirculatingSupply(marketCap: number, price: number) {
```
Comment (mateuszradomski) [inline packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts:232]: Document this function, what it does and how it achieves that

**Review by mateuszradomski — APPROVED** [review]
lgtm, GJ!

File: packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts (hunk 4 above)
Comment (antooni) [reply packages/backend/src/peripherals/coingecko/CoingeckoQueryService.ts:40]: they are needed for the await expression

**Pull request merged**